common-utils/src/main/java/eu/arrowhead/common/init/ApplicationInitListener.java

```diff
@@ -167,9 +199,98 @@ private void obtainKeys(final KeyStore keyStore) {
 	}
 
 	//-------------------------------------------------------------------------------------------------
-	private void registerToServiceRegistry() {
-		// TODO implement
-		// register system and its services
+	private void registerToServiceRegistry() throws InterruptedException {
+		logger.debug("registerToServiceRegistry started...");
+
+		if (skipRegistration()) {
+			return;
+		}
+
+		checkServiceRegistryConnection(sysInfo.isSslEnabled(), MAX_NUMBER_OF_SERVICEREGISTRY_CONNECTION_RETRIES, WAITING_PERIOD_BETWEEN_RETRIES_IN_SECONDS);
+
+		// revoke system (if any)
+		arrowheadHttpService.consumeService(Constants.SERVICE_DEF_SYSTEM_DISCOVERY, Constants.SERVICE_OP_REVOKE, Void.class);
+
+		// register system
+		final SystemModel model = sysInfo.getSystemModel();
+		final List<AddressDTO> addresses = model.addresses()
+				.stream()
+				.map(a -> new AddressDTO(a.type().name(), a.address()))
+				.collect(Collectors.toList());
+		final SystemRegisterRequestDTO payload = new SystemRegisterRequestDTO(model.metadata(), model.version(), addresses, model.deviceName());
+		arrowheadHttpService.consumeService(Constants.SERVICE_DEF_SYSTEM_DISCOVERY, Constants.SERVICE_OP_REGISTER, SystemResponseDTO.class, payload);
+
+		// register services
+		for (final ServiceModel serviceModel : sysInfo.getServices()) {
+			registerService(serviceModel);
+		}
+
+		logger.info("System {} published {} service(s).", sysInfo.getSystemName(), registeredServices.size());
 	}
 
+	//-------------------------------------------------------------------------------------------------
+	private boolean skipRegistration() {
+		logger.debug("skipRegistration started...");
+
+		return standaloneMode || Constants.SYS_NAME_SERVICE_REGISTRY.equals(sysInfo.getSystemName());
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	private void checkServiceRegistryConnection(final boolean secure, final int retries, final int period) throws InterruptedException {
+		logger.debug("checkServiceRegistryConnection started...");
+
+		final String templateName = secure ? Constants.GENERIC_HTTPS_INTERFACE_TEMPLATE_NAME : Constants.GENERIC_HTTP_INTERFACE_TEMPLATE_NAME;
+		for (int i = 0; i <= retries; ++i) {
```

Comment:
Why not `i < retries` ?

Reply:
Because the name is retries. If retries value is 3, you have to try 4 times. (1 original and 3 retries)